metadata/metadata.c

```diff
@@ -10,94 +12,104 @@
+        return;
+    }
 
-void metadata_storage_backend_register(MetadataBackend *backend) {
     assert(backend->name);
-    SLIST_INSERT_HEAD(&g_storage_backends, backend, slist);
+
+    MetadataBackendContext *context = malloc(sizeof(MetadataBackendContext));
+    context->was_already_initialized = false;
+    context->is_safe_to_reuse = is_safe_to_reuse;
+    context->backend = backend;
+
+    SLIST_INSERT_HEAD(&g_storage_backends, context, slist);
 }
 
-int metadata_persist(Metadata *metadata) {
+int metadata_persist(HyperwarpMetadata *metadata) {
     int ret = 0;
 
-    if (selected_backend == NULL) {
+    if (metadata->backend_context == NULL) {
         return 1;
     }
 
-    selected_backend->persist(metadata);
-
-    return ret;
-}
-
-Metadata *metadata_load() {
-    if (selected_backend == NULL) {
-        return NULL;
-    }
-
-    return selected_backend->load();
+    return metadata->backend_context->backend->persist(metadata->metadata);
 }
 
-int metadata_backend_initialize() {
+int metadata_backend_initialize(HyperwarpMetadata *metadata) {
     int ret = 0;
 
-    if (selected_backend == NULL) {
+    if (metadata->backend_context == NULL) {
         return 1;
     }
 
-    if ((ret = selected_backend->initialize()) != 0) {
+    if ((metadata->backend_context->is_safe_to_reuse == false) &&
+            (metadata->backend_context->was_already_initialized)) {
+        fprintf(stderr, "The selected module cannot be reinitialized\n");
+        return 1;
+    }
+
+    if ((ret = metadata->backend_context->backend->initialize(metadata->backend_context->was_already_initialized)) != 0) {
         return ret;
     }
 
+    metadata->backend_context->was_already_initialized = true;
+
     return ret;
 }
 
 
-int metadata_backend_finalize() {
+int metadata_backend_finalize(HyperwarpMetadata *metadata) {
     int ret = 0;
 
-    if (selected_backend == NULL) {
-        return 1;
+    if (metadata->backend_context == NULL) {
+        // if there is no open backend then there is nothing to do
+        return 0;
     }
 
-    if ((ret = selected_backend->finalize()) != 0) {
+    if ((ret = metadata->backend_context->backend->finalize()) != 0) {
         return ret;
     }
 
     return ret;
 }
 
-MetadataBackend *get_metadata_backend_by_name(const char* name) {
-    MetadataBackend *backend;
+MetadataBackendContext *get_metadata_backend_context_by_name(const char* name) {
+    MetadataBackendContext *context;
 
-    SLIST_FOREACH(backend, &g_storage_backends, slist) {
-        if (strcmp(backend->name, name) == 0) {
-            return backend;
+    SLIST_FOREACH(context, &g_storage_backends, slist) {
+        assert(context->backend != NULL);
+        if (strcmp(context->backend->name, name) == 0) {
+            return context;
         }
     }
 
     return NULL;
 }
 
-int use_metadata_storage_backend(const char *name) {
-    if (selected_backend != NULL) {
+int use_metadata_storage_backend(HyperwarpMetadata *metadata, const char *name) {
```

Comment:
what do you think of naming the parameter `backend_name` just to be super explicit about what is expected?